### DISCUSSION TITLE
Thank you for making this

### DISCUSSION THREAD

**Discussion by LucidFir** [erew123/alltalk_tts#131] [General]
This is such a well made, easy to install, perfectly documented, just fantastic all around thing. Thank you. 

I have installed (or tried and failed to) just about every AI TTS that exists and so yeah. Good job :)

**Comment by erew123** [comment]
Hi @LucidFir 

Thanks! I really appreciate you saying that! :)

Always nice to hear.

**Comment by LucidFir** [comment]
If any error codes or anything would be useful, I set up a loop to convert
text files to audio. About 50 files later it started giving me 400 errors I
can't find a way past

**Reply by erew123** [reply]
@LucidFir There are limitations as to what the XTTS model can handle at one time. It preferably only handles 250 character chunks. I have no idea how much or what kind of text you are sending over or how you assess when the model has completed the last block of text, so perhaps the issue lays somewhere within that type of area.

Lets make a hypothetical situation here that your text files are 500 words long and have 30 sentences in them. 

You may look to have your script:

- open your text file,
- break the text into chunks (typically full sentences so you don't break the spoken flow of a sentence. Multiple sentences are fine as long as you aren't going beyond 250 characters)
- generate each chunk, gathering the returned wav file into a array of files.
- when all chunks for the text file have been gathered into the array, join the audio together into one file.
- move onto the next file.

This would certainly provide breaks in the process that ensure you are not sending too much to the model or overloading it. Its pretty much the process that the TTS generator uses to ensure that it can generate unlimited text. https://github.com/erew123/alltalk_tts?tab=readme-ov-file#-alltalk-tts-generator

**Reply by erew123** [reply]
It would be **something** like the below. This is NOT a tested/working script. Im just purely giving an example of the kind of path required to make something to do the task.

```
import os
import requests
from pydub import AudioSegment

def split_text_into_chunks(text, limit=250):
    """
    Splits text into chunks, each with a maximum length of 'limit'.
    Attempts to split by sentence for natural breaks.
    """
    sentences = text.split('.')
    chunks = []
    current_chunk = ""
    
    for sentence in sentences:
        if len(current_chunk) + len(sentence) <= limit:
            current_chunk += sentence + '.'
        else:
            chunks.append(current_chunk)
            current_chunk = sentence + '.'
    if current_chunk:
        chunks.append(current_chunk)
    
    return chunks

def generate_audio_for_text(text_chunk, output_file_prefix):
    """
    Generates audio for a given text chunk and returns the local path of the generated audio file,
    including all the necessary parameters for the TTS API.
    """
    response = requests.post("http://127.0.0.1:7851/api/tts-generate", data={
        "text_input": text_chunk,
        "text_filtering": "standard",
        "character_voice_gen": "female_01.wav",  # The voice to use
        "narrator_enabled": "false",
        "narrator_voice_gen": "male_01.wav",  # Not used, but required by the API
        "text_not_inside": "character",
        "language": "en",
        "output_file_name": output_file_prefix,
        "output_file_timestamp": "true",
        "autoplay": "false",
        "autoplay_volume": "0.8",
    })
    
    if response.status_code == 200:
        data = response.json()
        if data['status'] == 'generate-success':
            return data['output_file_path']  # Assuming this is a local path
    return None

def process_text_file(file_path):
    """
    Processes a single text file, generates audio for each chunk, and merges them into a single audio file.
    """
    with open(file_path, 'r') as file:
        text = file.read()
        
    chunks = split_text_into_chunks(text)
    combined_audio = None
    
    for i, chunk in enumerate(chunks):
        print(f"Processing chunk {i+1}/{len(chunks)}...")
        audio_path = generate_audio_for_text(chunk, "output_chunk")
        if audio_path:
            audio_segment = AudioSegment.from_wav(audio_path)
            combined_audio = audio_segment if combined_audio is None else combined_audio + audio_segment
    
    if combined_audio:
        combined_file_name = f"{os.path.splitext(file_path)[0]}_combined.wav"
        combined_audio.export(combined_file_name, format="wav")
        print(f"Combined audio saved as: {combined_file_name}")

# Example usage:
directory_path = '/path/to/your/text/files'
for filename in os.listdir(directory_path):
    if filename.endswith(".txt"):
        print(f"Processing file: {filename}")
        process_text_file(os.path.join(directory_path, filename))
```
